### PR TITLE
TraitName should be unique in application webhook

### PR DESCRIPTION
**Pull request by Cweiping**
Fixed: https://github.com/oam-dev/kubevela/issues/1093

### CODE REVIEW

**Review by wonderflow — APPROVED** [review]
Thanks for contributions. Can you help write a test for it?

**Comment by Cweiping** [comment]
> Thanks for contributions. Can you help write a test for it?

Done

File: pkg/webhook/core.oam.dev/v1alpha2/application/helper.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	klog.Info("validate trait name conflicts ", "app name:", v.app.Name)
+	allErrs := make([]error, 0)
+	allTraitName := make(map[string]string)
+	for _, component := range v.app.Spec.Components {
+		for _, trait := range component.Traits {
+			if compName, ok := allTraitName[trait.Name]; ok {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), v.app.Spec,
+					fmt.Sprintf(errFmtTraitNameConflict, compName, trait.Name, component.Name, trait.Name)))
+			} else {
+				allTraitName[trait.Name] = component.Name
```
Comment (wonderflow) [inline pkg/webhook/core.oam.dev/v1alpha2/application/helper.go:46]: trait only conflicts within one component. We can use same trait in different components.

**Review by wonderflow — CHANGES_REQUESTED** [review]
Can you add a e2e test case for it? Thanks

File: pkg/webhook/core.oam.dev/v1alpha2/application/helper.go (hunk 1 above)
Comment (Cweiping) [reply pkg/webhook/core.oam.dev/v1alpha2/application/helper.go]: Done.

**Comment by Cweiping** [comment]
> Can you add a e2e test case for it? Thanks

Done.Please review again.

File: pkg/webhook/core.oam.dev/v1alpha2/application/helper.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/klog"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha2"
+)
+
+const (
+	errFmtTraitNameConflict = "traits name (%q) conflict between of component %q"
```
Comment (wonderflow) [inline pkg/webhook/core.oam.dev/v1alpha2/application/helper.go:14]:
```suggestion
	errFmtTraitNameConflict = "trait resources (%s) conflicts between trait(%s) and trait(%s) in component %s"
```

File: pkg/webhook/core.oam.dev/v1alpha2/application/helper.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	allErrs := make([]error, 0)
+	for _, component := range v.app.Spec.Components {
+		allTraitName := make(map[string]bool)
+		for _, trait := range component.Traits {
+			if _, ok := allTraitName[trait.Name]; ok {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), v.app.Spec,
+					fmt.Sprintf(errFmtTraitNameConflict, trait.Name, component.Name)))
+			} else {
+				allTraitName[trait.Name] = true
+			}
```
Comment (wonderflow) [inline pkg/webhook/core.oam.dev/v1alpha2/application/helper.go:47]: We should check the trait resource name not the trait name:

In the below example, the trait resource name is `service` and `ingress`. So you should parse the CUE template in this validation.

```
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: ingress
spec:
  extension:
    template: |
      parameter: {
        domain: string
        path: string
        exposePort: int
      }
      // trait template can have multiple outputs in one trait
      outputs: service: {
        apiVersion: "v1"
        kind: "Service"
        spec: {
          selector:
            app: context.name
          ports: [{
              port: parameter.exposePort
              targetPort: context.output.spec.template.spec.containers[0].ports[0].containerPort
            }]
        }
      }
      outputs: ingress: {
          apiVersion: "networking.k8s.io/v1beta1"
          kind: "Ingress"
          metadata:
            name: context.name
            labels: config: context.outputs.gameconfig.data.enemies
          spec: {
            rules: [{
              host: parameter.domain
              http: {
                paths: [{
                    path: parameter.path
                    backend: {
                      serviceName: context.name
                      servicePort: parameter.exposePort
                    }
                }]
              }
            }]
          }
      }
```

**Comment by wonderflow** [comment]
prefer #1159 , closing

Thanks for your contribution

**Pull request closed without merging**